rust/main/applications/hyperlane-application/src/operation_verifier.rs

```diff
@@ -0,0 +1,31 @@
+use async_trait::async_trait;
+
+use hyperlane_core::{ChainCommunicationError, HyperlaneMessage};
+
+/// Error which can be reported by verifier
+#[derive(Debug, thiserror::Error)]
+pub enum ApplicationOperationVerifierError {
+    /// Message is malformed
+    #[error("Malformed message: {0:?}")]
+    MalformedMessageError(HyperlaneMessage),
+    /// Chain communication error
+    #[error("Remote communication error: {0:?}")]
+    ChainCommunicationError(ChainCommunicationError),
+    /// Insufficient amount
+    #[error("Insufficient amount")]
+    InsufficientAmountError,
+    /// Unknown application
+    #[error("Unknown app context: {0:?}")]
+    UnknownApplicationError(String),
+}
+
+/// Trait to verify if operation is permitted for application context
+#[async_trait]
+pub trait ApplicationOperationVerifier: Send + Sync {
+    /// Checks if message is permitted for application context
+    async fn verify(
+        &self,
+        app_context: &Option<String>,
+        message: &HyperlaneMessage,
+    ) -> Result<(), ApplicationOperationVerifierError>;
```

Comment:
imo this is more of a labeling function that a verifying one. Even though at the application level we may treat these labels as errors, idiomatic would be to not rely on the returned Error type unless something bad happened in this function. If the AppVerifier isn't able to provide a label that's not an error, and neither is labelling that e.g. this message transfers zero value on cosmos

So rather than use `Result<(), Error>` and always rely on the error, I think we should use `Option<AppSpecificInfo>` or similar. Being able to apply a label is not an error as far as this method is concerned, and being unable to apply a label is fine and we retun `None`

Reply:
Implemented